File: src/main/java/nz/zephire/exceptions/exceptions/ExceptionManager.java
package nz.zephire.exceptions.exceptions;

import nz.zephire.exceptions.StateConfig;
import org.joda.time.LocalDate;
import org.joda.time.LocalTime;

public class ExceptionManager {

//    LocalTime a = new LocalTime(5,30);
//    LocalTime b = new LocalTime(6, 0);
//    Period p = new Period(a, b);
//    p.toStandardDuration().getStandardMinutes(); // TODO: length of exception/differences

    private final StateConfig config;
    private final LocalDate date = LocalDate.now();

    private static ExceptionManager inst;

    public static ExceptionManager inst() {
        if(inst == null) return new ExceptionManager(StateConfig.loadConfig());

        return inst;
    }

    private ExceptionManager(StateConfig config) {
        this.config = config;

        inst = this;
    }

    public Exception getNextPreApproved() {
        LocalTime now = LocalTime.now();

        Exception next = null;
        for(PreApprovedException e : config.getPreApproved()) {
            if(e.isComplete()) continue;

            if(next == null) {;
                next = e;
            } else {
                if(next.getStart().compareTo(e.getStart()) > 0) { // >0 == earliest is later than e
                    next = e;
                }
            }
        }

        return next;
    }

    public void addException(Exception exception) {
        if(exception instanceof PreApprovedException) {
            config.addPreApproved((PreApprovedException) exception);
            return;
        }

        config.addException(exception);
    }

    // Simple Getters

    public LocalDate getDate() {
        return date;
    }

    public StateConfig getConfig() {
        return config;
    }
}

File: src/main/java/nz/zephire/exceptions/exceptions/ExceptionManager.kt
package nz.zephire.exceptions.exceptions

import nz.zephire.exceptions.StateConfig
import nz.zephire.exceptions.loadConfig

private var inst: ExceptionManager? = null

fun inst(): ExceptionManager {
    return inst ?: ExceptionManager(loadConfig())
}

class ExceptionManager internal constructor(val config: StateConfig) {

    val nextPreApproved: Exception?
        get() {
            var next: Exception? = null
            for (e in config.rostered) {
                if (e.isComplete) continue

                if (next == null) {
                    next = e
                } else {
                    if (e.start < next.start) {
                        next = e
                    }
                }
            }

            return next
        }

    /**
     * Fetches and gives the current adherence percentage
     * @return Adherence percentage between 0-100
     */
    val adherencePercent: Int
        get() = 95 //TODO: Fetch/Cache

    init {
        inst = this
    }

    fun addException(exception: Exception) {
        if (exception is RosteredException) {
            config.addRosteredException(exception)
            return
        }

        config.addException(exception)
    }
}
